File: src/app/core/gitWorkspace.ts
export type GitFileChange = {
  path: string;
  status: string;
  staged: boolean;
  unstaged: boolean;
};

export type GitWorkspaceStatus = {
  repo_path: string;
  branch: string;
  upstream: string | null;
  clean: boolean;
  changes: GitFileChange[];
  ahead: number;
  behind: number;
};

export type DiffFile = {
  path: string;
  status: string;
  additions: number;
  deletions: number;
  hunks: string[];
};

export type DiffView = {
  repo_path: string;
  files: DiffFile[];
  raw: string;
  additions: number;
  deletions: number;
};

export type CommitProposal = {
  message: string;
  paths: string[];
  summary: string;
  risk: string;
};

export type CommitResult = {
  success: boolean;
  proposal: CommitProposal;
  commit_sha: string | null;
};

export type PullRequestDraft = {
  title: string;
  body: string;
  head_branch: string;
  base_branch: string;
  changed_files: string[];
  diff_summary: string;
  risks: string[];
  linked_tasks: string[];
};

export async function gitStatus(repoPath = "."): Promise<GitWorkspaceStatus> {
  return postGit<GitWorkspaceStatus>("/status", { repo_path: repoPath });
}

export async function createGitBranch(repoPath: string, branch: string, baseBranch?: string) {
  return postGit("/branches", { repo_path: repoPath, branch, base_branch: baseBranch });
}

export async function gitDiff(repoPath = ".", paths: string[] = [], staged = false): Promise<DiffView> {
  return postGit<DiffView>("/diff", { repo_path: repoPath, paths, staged });
}

export async function generateCommitProposal(
  repoPath: string,
  description: string,
  paths: string[] = [],
  kind = "feat",
  scope?: string,
): Promise<CommitProposal> {
  return postGit<CommitProposal>("/commits/proposal", {
    repo_path: repoPath,
    description,
    paths,
    kind,
    scope,
  });
}

export async function commitGitChanges(repoPath: string, proposal: CommitProposal): Promise<CommitResult> {
  return postGit<CommitResult>("/commits", {
    repo_path: repoPath,
    message: proposal.message,
    paths: proposal.paths,
    summary: proposal.summary,
    risk: proposal.risk,
  });
}

export async function preparePullRequestDraft(
  repoPath: string,
  description: string,
  options: { title?: string; baseBranch?: string; linkedTasks?: string[] } = {},
): Promise<PullRequestDraft> {
  return postGit<PullRequestDraft>("/pull-request/draft", {
    repo_path: repoPath,
    title: options.title,
    description,
    base_branch: options.baseBranch ?? "main",
    linked_tasks: options.linkedTasks ?? [],
  });
}

async function postGit<T>(path: string, payload: Record<string, unknown>): Promise<T> {
  const response = await fetch(`/api/git${path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    const error = await response.text();
    throw new Error(error || `Git workspace request failed: ${path}`);
  }
  return (await response.json()) as T;
}

File: backend/api/routes/git_workspace.py
from pathlib import Path

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from anubis.workspace import CommitProposal, GitWorkspace, GitWorkspaceConfig, GitWorkspaceError
from backend.core.config import settings


router = APIRouter()


class RepoRequest(BaseModel):
    repo_path: str = "."


class CreateBranchRequest(RepoRequest):
    branch: str = Field(min_length=1)
    base_branch: str | None = None


class DiffRequest(RepoRequest):
    paths: list[str] = Field(default_factory=list)
    staged: bool = False


class GenerateCommitRequest(RepoRequest):
    description: str = Field(min_length=1)
    paths: list[str] = Field(default_factory=list)
    kind: str = "feat"
    scope: str | None = None


class CommitRequest(RepoRequest):
    message: str = Field(min_length=1)
    paths: list[str] = Field(default_factory=list)
    summary: str = ""
    risk: str = "unknown"


class PreparePRRequest(RepoRequest):
    title: str | None = None
    description: str = ""
    base_branch: str = "main"
    linked_tasks: list[str] = Field(default_factory=list)


def workspace(repo_path: str) -> GitWorkspace:
    try:
        return GitWorkspace(
            repo_path,
            config=GitWorkspaceConfig(root=Path(settings.project_root), timeout_seconds=settings.tool_timeout_seconds),
        )
    except GitWorkspaceError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/status")
def status(payload: RepoRequest) -> dict[str, object]:
    return workspace(payload.repo_path).status().to_dict()


@router.post("/branches")
def create_branch(payload: CreateBranchRequest) -> dict[str, object]:
    return workspace(payload.repo_path).create_branch(payload.branch, base_branch=payload.base_branch).to_dict()


@router.post("/diff")
def diff(payload: DiffRequest) -> dict[str, object]:
    return workspace(payload.repo_path).diff(paths=tuple(payload.paths), staged=payload.staged).to_dict()


@router.post("/commits/proposal")
def generate_commit(payload: GenerateCommitRequest) -> dict[str, object]:
    return workspace(payload.repo_path).generate_commit(
        description=payload.description,
        paths=tuple(payload.paths),
        kind=payload.kind,
        scope=payload.scope,
    ).to_dict()


@router.post("/commits")
def commit(payload: CommitRequest) -> dict[str, object]:
    proposal = CommitProposal(
        message=payload.message,
        paths=tuple(payload.paths),
        summary=payload.summary,
        risk=payload.risk,
    )
    return workspace(payload.repo_path).commit(proposal).to_dict()


@router.post("/pull-request/draft")
def prepare_pr(payload: PreparePRRequest) -> dict[str, object]:
    return workspace(payload.repo_path).prepare_pr(
        title=payload.title,
        description=payload.description,
        base_branch=payload.base_branch,
        linked_tasks=tuple(payload.linked_tasks),
    ).to_dict()
